File: C++/myfleetsheets_api.cpp
#include "nlohmann/json.hpp" // this is the single include file https://github.com/nlohmann/json/blob/develop/single_include/nlohmann/json.hpp
#include <string>
#include <iostream>
#include <fstream>
#include <stdio.h>

// for convenience
using json = nlohmann::json;
using namespace std;

struct MFS_Credentials {
	string email;
	string key;
};

MFS_Credentials get_creds(string filename) {
	ifstream t;
    t.open(filename);
    string whole, line;
    while(!t.eof()){
        getline(t,line);
        whole += line;
    }
    t.close();

    json file_json = json::parse(whole);
    MFS_Credentials creds = MFS_Credentials();
    creds.email = file_json["email"];
    creds.key = file_json["key"];
    return creds;
}

string exec(const char* cmd) {
    char buffer[128];
    string result = "";
    FILE* pipe = popen(cmd, "r");
    if (!pipe) throw runtime_error("popen() failed!");
    try {
        while (fgets(buffer, sizeof buffer, pipe) != NULL) {
            result += buffer;
        }
    } catch (...) {
        pclose(pipe);
        throw;
    }
    pclose(pipe);
    return result;
}

string APIrequest(string creds_filename = "", string email = "", string key = "", string since = "", string make = "", string model = "", 
	              string serial = "", string year = "", string max_price = "", string min_price = "", string source = "") {

	string command = "curl -ksb -H \"Accept: application/json\" \"https://api.myfleetsheets.com/api/planes/?";

	if (creds_filename != "") {
		if (email != "" || key != "") {
			printf("If a credentials file is provided, there is no need to provide email and API key separately");
			exit(EXIT_FAILURE);
		}
		MFS_Credentials creds = get_creds(creds_filename);
		email = creds.email; key = creds.key;
	}

	string fields[10]     = { email,   key,   since,   make,   model,   serial,   year,   max_price,   min_price,   source};
	string fieldnames[10] = {"email", "key", "since", "make", "model", "serial", "year", "max_price", "min_price", "source"};
	for (int i = 0; i < 10; i++) {
		if (fields[i] != "") {
			command += fieldnames[i] + "=" + fields[i] + "&";
		}
	}

	command += "\"";
	// printf("%s\n", command.c_str());

	return exec(command.c_str());	
}

int main(int argc, char* argv[]) {
	string creds_filename = "", email = "", key = "", since = "", make = "", model = "", serial = "", year = "",
	       max_price = "", min_price = "", source = "";

	int argi = 1;
	while (argi < argc) {
		string flag = argv[argi];
		if (argi + 1 == argc) { printf("Argument must have value: %s\n", flag.c_str()); exit(EXIT_FAILURE); }
		string value = argv[argi+1];
		argi += 2;

		if      (flag == "-f")          { creds_filename = value; }
		else if (flag == "-e")          { email          = value; }
		else if (flag == "-k")          { key            = value; }
		else if (flag == "--since")     { since          = value; }
		else if (flag == "--make")      { make           = value; }
		else if (flag == "--model")     { model          = value; }
		else if (flag == "--serial")    { serial         = value; }
		else if (flag == "--year")      { year           = value; }
		else if (flag == "--max_price") { max_price      = value; }
		else if (flag == "--min_price") { min_price      = value; }
		else if (flag == "--source")    { source         = value; }
		else { printf("Unknown argument: %s\n", flag.c_str()); exit(EXIT_FAILURE); }
	}

	string http_result = APIrequest(creds_filename, email, key, since, make, model, serial, year, max_price, min_price, source);
	printf("%s\n", http_result.c_str());
	return 0;
}

File: Python/myfleetsheets_api.py
import json
import sys
import argparse
from urllib import request as ur
from urllib import parse as up
import ssl

parser = argparse.ArgumentParser()
parser.add_argument('-f', metavar="file", type=str, nargs = "?",
                    help = "credentials JSON file")
parser.add_argument('-e', metavar="email", type=str, nargs = "?",
                    help = "your email address")
parser.add_argument('-k', metavar="file", type=str, nargs = "?",
                    help = "your API key")
parser.add_argument('--since', metavar="delta", type=str, nargs = "?",
                    help = "a time expression representing how far in the past to look, e.g., 6h, 7d, 2w")
parser.add_argument('--make', metavar="make1,make2,...", type=str, nargs = "?",
                    help = "a make, or comma separated list of makes")
parser.add_argument('--model', metavar="model1,model2,...", type=str, nargs = "?",
                    help = "a model, or comma-separated list of models")
parser.add_argument('--serial', metavar="serial1,serial2,...", type=str, nargs = "?",
                    help = "a serial, or comma-separated list of serials")
parser.add_argument('--year', metavar="year1-year2", type=str, nargs = "?",
                    help = "a year, a comma-separated list of years, or a range of years")
parser.add_argument('--max_price', metavar="price", type=str, nargs = "?",
                    help = "the maximum asking price")
parser.add_argument('--min_price', metavar="price", type=str, nargs = "?",
                    help = "the minimum asking price; setting min_price excludes planes with unknown or non-numeric prices")
parser.add_argument('--source', metavar="source1,source2,...", type=str, nargs = "?",
                    help = "a source, or comma-separated list of sources")

API_PARAM_FIELDS = ["since","make","model","serial","year","max_price","min_price","source"]

API_BASE_URL = "https://api.myfleetsheets.com/api/planes/?"


def APIrequest(**kwargs):
    if kwargs.get("f"):
        if kwargs.get("e") or kwargs.get("k"):
            raise ValueError("If a credentials file is provided, there is no need to provide email and API key separately")
            
        try:
            with open(kwargs["f"],"r",encoding="utf-8") as fh:
                creds = json.load(fh)
        except FileNotFoundError:
            raise ValueError("File not found: " + kwargs["f"])
            
        try:
            email = creds["email"]
            key = creds["key"]
        except:
            raise ValueError("Invalid credentials file")

    else:
        if not (kwargs.get("e") and kwargs.get("k")):
            raise ValueError("Please pass your email address and API key")

        email = kwargs["e"]
        key = kwargs["k"]

    url = API_BASE_URL + "email=%s&key=%s&" % (email, key) 

    for kw, arg in kwargs.items():
        if kw in API_PARAM_FIELDS and arg is not None:
            if type(arg) is list:
                arg = ",".join(map(lambda x: str(x), arg))
                
            url += "%s=%s&" % (kw, up.quote(str(arg)))

    response = ur.urlopen(url)

    raw_json = response.read()

    try:
        return json.loads(raw_json)
    except json.decoder.JSONDecodeError as e:
        raise ValueError(raw_json)

if __name__ == "__main__":
    kwargs = vars(parser.parse_args())

    try:
        listing_json = APIrequest(**kwargs)
    except ValueError as e:
        print(e)
        sys.exit()

    formatted_json = json.dumps(listing_json, indent=4, sort_keys = True)
    print(formatted_json)
